### ISSUE TITLE
Some way to have circular bounds?

### ISSUE THREAD

**Issue by entropylost** [andreesteve/voronoice#7]
I would like to have a circular boundary for my voronoi diagram, or at least a close approximation using polygons. This could be done by replacing the bounding-box with a `Polygon` type.

**Comment by andreesteve** [comment]
This sounds very interesting! In principle, as you pointed out, the bounding box would have to be replaced with a more generic construct. The code does have some assumptions right now that the constraint is a rectangle so the clipping logic would have to be changed as well. I wonder if all assumptions currently made for clipping on a box would apply to a generic convex polygon.

**Comment by entropylost** [comment]
Perhaps just assume that the points are generated within the polygon or bounding box but clip all the lines to it?

**Comment by SPuntte** [comment]
I realize this issue is over a year old, but what I've been doing in [my fork](https://github.com/SPuntte/voronoice) recently seems closely related. I can't honestly claim to understand all the details of the diagram generation logic in this library yet but so far it certainly seems to generalize pretty straightforwardly to arbitrary convex polygons.

@andreesteve, maybe a separate issue would be appropriate if you're interested in collaborating on this?

**Comment by andreesteve** [comment]
Hi @SPuntte - thanks for looking into this ! I believe indeed generalizing the bounding box into a convex polygon and the clipping algorithm into [Sutherland–Hodgman](https://en.wikipedia.org/wiki/Sutherland%E2%80%93Hodgman_algorithm) is a good way to go, and would be a great addition to the library.

I think such a change would be welcomed to the library. Especially if we can bring the generalization with similar benchmark results when the polygon is a bounding box.

We can use this issue or file a new one. No preference from my side. If we file a new one, I'd duplicate this issue against that. Let me know your thoughts.

**Comment by Antoshidza** [comment]
I think one of the best way would be not to replace `BoundingBox` but to implement `Bounds` trait, so any of us can implement bounding logic as simple as they like. In this case solution will have basic impl as `BoundingBox` and maybe if someone can still do this some `PolygonBound`

**Comment by Dtrasla** [comment]
Adding to this, found a thread this week as I wanted to make use of a Polygonal bounding box using Delaunay triangulation, I might open up a PR once I'm done with a personal project

https://stackoverflow.com/questions/23823345/how-to-construct-a-voronoi-diagram-inside-a-polygon 
https://en.wikipedia.org/wiki/Bowyer%E2%80%93Watson_algorithm

**Comment by andreesteve** [comment]
@Dtrasla, I'd welcome a PR that adds supports to a more generic boundary! Sharing a few suggestions on in case it helps. I see two options, the easier one and the harder one :)

The harder approach would be, as @Antoshidza suggested, to generalize the bounding box as a Boundary trait, change the current interface to take a generic implementation of such trait, and generalize the logic in the cell builder. The harder part here is to figure out the right representation of voronoi cells, especially the ones in the hull that will be clipped. Currently the library assumes those are closed convex polygons, so there might be other implications. I'd imagine that we would need to figure out those implications first.

I believe there is a much easier way to go about this, that probably achieves a similar, if not indistinguishable, result. The easier approach is to evolve the bounding box implementation into a regular convex polygon. You can keep the same algorithm, orchestration in ```cell_builder.rs``` and overall the theme here would be iterating over the edges when projecting rays into the boundary and generalizing the corner checks. Once that is done, one could get a reasonable voronoi with circular bounds by bounding it within a regular convex polygon with a high edge count.
